File: apps/client-solid/src/lib/describe.ts
import type { HookEvent } from './types';

export interface EventDescription {
  primary: string;
  secondary?: string;
}

export function sortedPayloadEntries(
  payload: Record<string, unknown> | null | undefined,
): Array<[string, unknown]> {
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) return [];
  return Object.entries(payload).sort(([left], [right]) => left.localeCompare(right));
}

export function isPayloadContainer(value: unknown): value is Record<string, unknown> | unknown[] {
  return Array.isArray(value) || (typeof value === 'object' && value !== null);
}

export function formatPayloadScalar(value: unknown): string {
  if (value === null) return 'null';
  if (typeof value === 'string') return value;
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  if (typeof value === 'undefined') return 'undefined';
  try {
    return JSON.stringify(value);
  } catch {
    return String(value);
  }
}

export function simpleEventKindLabel(type: string): string {
  switch (type) {
    case 'UserPromptSubmit':
      return 'You';
    case 'AssistantResponse':
      return 'Reply';
    case 'PreToolUse':
      return 'Doing';
    case 'PostToolUse':
      return 'Done';
    case 'PostToolUseFailure':
      return 'Failed';
    case 'SubagentStart':
      return 'Delegated';
    case 'SubagentStop':
      return 'Completed';
    default:
      return 'Update';
  }
}

export function mergeSimpleRowSummary(event: HookEvent, maxLen = 120): string {
  const description = describeEventDescription(event);
  const primary = description.primary.trim();
  const secondary = description.secondary?.trim();
  if (!secondary || primary.toLowerCase().includes(secondary.toLowerCase())) {
    return truncate(primary, maxLen);
  }
  const joined = `${primary} - ${secondary}`;
  if (joined.length <= maxLen) return joined;
  return truncate(joined, maxLen);
}

/** Trust signal: observed vs managed acquisition; unknown or absent values yield no label */
export function formatAcquisitionModeLabel(mode?: string | null): string | undefined {
  if (typeof mode !== 'string') return undefined;
  const normalized = mode
    .trim()
    .toLowerCase()
    .replace(/[_-]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
  if (!normalized) return undefined;

  const first = normalized.split(' ')[0];
  if (!first || first === 'unknown') return undefined;

  if (first === 'observed') return 'Observed';
  if (first === 'managed') return 'Managed';

  return undefined;
}

export function formatRuntimeLabel(label?: string | null): string | undefined {
  if (typeof label !== 'string') return undefined;
  const normalized = label.trim().toLowerCase().replace(/[_-]+/g, ' ');
  if (!normalized) return undefined;

  if (normalized.includes('claude')) return 'Claude';
  if (normalized.includes('codex')) return 'Codex';
  if (normalized.includes('gemini')) return 'Gemini';
  if (normalized.includes('cursor')) return 'Cursor';

  return normalized
    .split(/\s+/)
    .filter(Boolean)
    .map((part) => part[0].toUpperCase() + part.slice(1))
    .join(' ');
}

/** One-line description of an event, optimized for operator scanning */
export function describeEvent(event: HookEvent): string {
  return describeEventDescription(event).primary;
}

/** Secondary technical detail for an event, when available */
export function describeEventDetail(event: HookEvent): string | undefined {
  return describeEventDescription(event).secondary;
}

/** Plain-language description plus secondary detail for the event */
export function describeEventDescription(event: HookEvent): EventDescription {
  const type = event.hook_event_type;
  const p = event.payload || {};
  const runtime = formatRuntimeLabel(p.runtime_label);
  const toolName = p.tool_name || 'unknown';
  const toolInput = p.tool_input || {};

  switch (type) {
    case 'PreToolUse': {
      const command = extractCommand(toolInput);
      if ((toolName === 'Bash' || toolName === 'exec_command' || toolName === 'Shell') && command) {
        const workdir = typeof toolInput?.workdir === 'string' ? basename(toolInput.workdir) : null;
        if (workdir) {
          return {
            primary: `Running a task in ${workdir}`,
            secondary: composeRuntimeDetail(runtime, command),
          };
        }
        return {
          primary: 'Running a task',
          secondary: composeRuntimeDetail(runtime, command),
        };
      }
      const fileTarget = extractFileTarget(toolInput);
      if (['Read', 'Edit', 'Write'].includes(toolName) && fileTarget) {
        const verb = toolName === 'Read' ? 'Reading a file' : toolName === 'Edit' ? 'Editing a file' : 'Writing a file';
        return {
          primary: verb,
          secondary: composeRuntimeDetail(runtime, fileTarget),
        };
      }
      if (toolName === 'apply_patch') {
        const patch = typeof toolInput?.patch === 'string'
          ? toolInput.patch
          : typeof toolInput?.input === 'string'
            ? toolInput.input
            : '';
        const target = extractPatchedFile(patch);
        return {
          primary: 'Updating files',
          secondary: composeRuntimeDetail(runtime, target),
        };
      }
      if (['Grep', 'Glob'].includes(toolName) && toolInput?.pattern) {
        return {
          primary: 'Searching for matches',
          secondary: composeRuntimeDetail(runtime, toolInput.pattern),
        };
      }
      if (toolName === 'Agent') {
        const description = toolInput?.description;
        return {
          primary: 'Handing off work to another agent',
          secondary: composeRuntimeDetail(runtime, description ? truncate(description, 80) : undefined),
        };
      }
      if (toolName === 'send_input' && toolInput?.message) {
        return {
          primary: 'Sending a message',
          secondary: composeRuntimeDetail(runtime, truncate(toolInput.message, 80)),
        };
      }
      if (toolName === 'wait_agent') {
        return {
          primary: 'Waiting for a helper agent to finish',
          secondary: composeRuntimeDetail(runtime),
        };
      }
      if (toolName === 'TodoWrite') {
        return {
          primary: 'Updating the plan',
          secondary: composeRuntimeDetail(runtime),
        };
      }
      return {
        primary: toolName !== 'unknown' ? `Using ${friendlyToolName(toolName)}` : 'Working on your request',
        secondary: composeRuntimeDetail(runtime),
      };
    }
    case 'PostToolUse': {
      const contentPreview = extractContentPreview(p.content);
      if (toolName === 'exec_command' && contentPreview) {
        return {
          primary: 'Finished the command',
          secondary: composeRuntimeDetail(runtime, truncate(contentPreview, 80)),
        };
      }
      if (contentPreview) {
        return {
          primary: 'Finished using a tool',
          secondary: composeRuntimeDetail(runtime, truncate(contentPreview, 80)),
        };
      }
      return {
        primary: 'Step finished',
        secondary: composeRuntimeDetail(runtime),
      };
    }
    case 'PostToolUseFailure': {
      const contentPreview = extractContentPreview(p.content);
      if (contentPreview) {
        return {
          primary: toolName === 'exec_command' ? 'Command failed' : 'A tool call failed',
          secondary: composeRuntimeDetail(runtime, truncate(contentPreview, 80)),
        };
      }
      return {
        primary: toolName === 'exec_command' ? 'Command failed' : 'Step failed',
        secondary: composeRuntimeDetail(runtime),
      };
    }
    case 'SessionStart':
      return {
        primary: p.title ? `Watching ${truncate(p.title, 80)}` : 'Watching the session',
        secondary: composeRuntimeDetail(runtime),
      };
    case 'SessionEnd':
      return {
        primary: 'Session ended',
        secondary: composeRuntimeDetail(runtime),
      };
    case 'SubagentStart': {
      const rawAgentLabel =
        p.display_name || p.agent_name || p.agent_type || event.agent_name || event.agent_type || 'Agent';
      const agentLabel = rawAgentLabel === 'Agent' ? 'a helper agent' : rawAgentLabel;
      if (p.description) {
        return {
          primary: `Started ${agentLabel}`,
          secondary: composeRuntimeDetail(runtime, truncate(p.description, 80)),
        };
      }
      return {
        primary: `Started ${agentLabel}`,
        secondary: composeRuntimeDetail(runtime),
      };
    }
    case 'SubagentStop':
      return {
        primary: 'Helper agent finished',
        secondary: composeRuntimeDetail(runtime),
      };
    case 'UserPromptSubmit': {
      const prompt = p.prompt || p.message || '';
      return {
        primary: prompt ? `You asked: ${truncate(prompt, 72)}` : 'You asked for help',
        secondary: composeRuntimeDetail(runtime),
      };
    }
    case 'AssistantResponse': {
      const text = p.text || '';
      return {
        primary: text ? truncate(text, 90) : 'Shared an update',
        secondary: composeRuntimeDetail(runtime),
      };
    }
    case 'SessionTitleChanged':
      return {
        primary: p.title ? `Renamed the session to ${truncate(p.title, 80)}` : 'Renamed the session',
        secondary: composeRuntimeDetail(runtime, p.title ? p.title : undefined),
      };
    default:
      return {
        primary: 'Observed an event',
        secondary: composeRuntimeDetail(runtime, type || undefined),
      };
  }
}

function truncate(s: string, max: number): string {
  return s.length > max ? s.slice(0, max) + '...' : s;
}

function extractCommand(toolInput: Record<string, any>): string | null {
  const candidate = toolInput?.cmd ?? toolInput?.command;
  if (typeof candidate === 'string' && candidate.trim()) {
    return candidate.trim();
  }
  if (Array.isArray(candidate)) {
    const parts = candidate
      .map((part) => (typeof part === 'string' ? part.trim() : ''))
      .filter(Boolean);
    if (parts.length > 0) {
      return parts.join(' ');
    }
  }
  return null;
}

function extractFileTarget(toolInput: Record<string, any>): string | null {
  const candidate = toolInput?.file_path ?? toolInput?.path ?? toolInput?.file;
  if (typeof candidate !== 'string' || !candidate.trim()) {
    return null;
  }
  return candidate.split('/').slice(-3).join('/');
}

function extractPatchedFile(patch: string): string | null {
  const match = patch.match(/\*\*\* (?:Add|Update|Delete) File: (.+)/);
  if (!match) {
    return null;
  }
  return match[1].split('/').slice(-3).join('/');
}

function extractContentPreview(content: unknown): string | null {
  if (typeof content !== 'string') {
    return null;
  }
  const firstLine = content
    .split('\n')
    .map((line) => line.trim())
    .find(Boolean);
  return firstLine || null;
}

function basename(path: string): string | null {
  const parts = path.split('/').filter(Boolean);
  return parts.length > 0 ? parts[parts.length - 1] : null;
}

function composeRuntimeDetail(runtime: string | undefined, detail?: string | null): string | undefined {
  const trimmedDetail = detail?.trim();
  if (runtime && trimmedDetail) {
    return `${runtime} runtime · ${trimmedDetail}`;
  }
  if (runtime) {
    return `${runtime} runtime`;
  }
  return trimmedDetail || undefined;
}

function friendlyToolName(toolName: string): string {
  const map: Record<string, string> = {
    readfile: 'file read',
    read: 'file read',
    edit: 'file edit',
    write: 'file write',
    shell: 'terminal command',
    exec_command: 'terminal command',
    grep: 'search',
    glob: 'file search',
    todowrite: 'plan update',
    agent: 'helper delegation',
    subagent: 'helper delegation',
  };
  const normalized = toolName.trim().toLowerCase();
  return map[normalized] || toolName;
}

File: apps/daemon-rs/src/live_state/describe.rs
use crate::model::LegacyHookEvent;

use super::util::{
    basename_from_path, clean_summary_text, content_preview, payload_string, short_path, truncate,
};

fn as_non_empty(value: Option<&str>) -> Option<String> {
    value
        .map(str::trim)
        .filter(|text| !text.is_empty())
        .map(ToString::to_string)
}

fn mcp_tool_summary(tool_name: &str, tool_input: &serde_json::Value) -> Option<String> {
    let uri = as_non_empty(tool_input.get("uri").and_then(serde_json::Value::as_str));
    let server = as_non_empty(tool_input.get("server").and_then(serde_json::Value::as_str));

    match tool_name {
        "CallMcpTool" => {
            let (server_l, mcp_l) = crate::cursor_callmcp::extract_server_tool(tool_input);
            let server_label = if server_l.is_empty() {
                as_non_empty(tool_input.get("server").and_then(serde_json::Value::as_str))
                    .unwrap_or_else(|| "unknown-server".to_string())
            } else {
                server_l
            };
            let mcp_tool = if !mcp_l.is_empty() {
                Some(mcp_l)
            } else {
                as_non_empty(
                    tool_input
                        .get("toolName")
                        .and_then(serde_json::Value::as_str)
                        .or_else(|| tool_input.get("tool_name").and_then(serde_json::Value::as_str)),
                )
            };
            if let Some(mcp_tool_name) = mcp_tool {
                let server_lc = server_label.to_ascii_lowercase();
                if server_lc.contains("librarian") {
                    let action = if mcp_tool_name.starts_with("memory_") {
                        mcp_tool_name
                            .strip_prefix("memory_")
                            .unwrap_or(mcp_tool_name.as_str())
                            .replace('_', " ")
                    } else {
                        mcp_tool_name.replace('_', " ")
                    };
                    return Some(format!("Librarian: {action}"));
                }
                if server_lc == "ai-memory-brain"
                    || server_lc == "user-ai-memory-brain"
                    || server_lc.ends_with("ai-memory-brain")
                    || mcp_tool_name.starts_with("memory_")
                {
                    let memory_action = mcp_tool_name
                        .strip_prefix("memory_")
                        .unwrap_or(mcp_tool_name.as_str())
                        .replace('_', " ");
                    return Some(format!("Memory brain: {memory_action}"));
                }
                return Some(format!("Calling MCP {server_label}/{mcp_tool_name}"));
            }
            Some(format!("Calling MCP {server_label}"))
        }
        "FetchMcpResource" => {
            let server_label = server.unwrap_or_else(|| "unknown-server".to_string());
            if let Some(resource_uri) = uri {
                let short_uri = truncate(&resource_uri, 48);
                return Some(format!("Fetching MCP resource {server_label}: {short_uri}"));
            }
            Some(format!("Fetching MCP resource from {server_label}"))
        }
        "ListMcpResources" => {
            let server_label = server.unwrap_or_else(|| "all servers".to_string());
            Some(format!("Listing MCP resources for {server_label}"))
        }
        _ => None,
    }
}

pub(crate) fn describe_legacy_event(event: &LegacyHookEvent) -> String {
    let tool_name =
        payload_string(&event.payload, "tool_name").unwrap_or_else(|| "unknown".to_string());
    match event.hook_event_type.as_str() {
        "PreToolUse" => {
            if let Some(tool_input) = event.payload.get("tool_input") {
                if (tool_name == "Bash" || tool_name == "exec_command")
                    && extract_command(tool_input).is_some()
                {
                    let command = extract_command(tool_input).unwrap_or_default();
                    if let Some(workdir) = tool_input
                        .get("workdir")
                        .and_then(serde_json::Value::as_str)
                        .and_then(basename_from_path)
                    {
                        return format!("Running {} in {workdir}", truncate(&command, 64));
                    }
                    return format!("Running {}", truncate(&command, 72));
                }
                if ["Read", "Edit", "Write"].contains(&tool_name.as_str())
                    && extract_file_target(tool_input).is_some()
                {
                    let target = extract_file_target(tool_input).unwrap_or_default();
                    let verb = match tool_name.as_str() {
                        "Read" => "Reading",
                        "Edit" => "Editing",
                        _ => "Writing",
                    };
                    return format!("{verb} {target}");
                }
                if tool_name == "apply_patch" {
                    let patch = tool_input
                        .get("patch")
                        .and_then(serde_json::Value::as_str)
                        .or_else(|| tool_input.get("input").and_then(serde_json::Value::as_str))
                        .unwrap_or_default();
                    if let Some(target) = extract_patched_file(patch) {
                        return format!("Patching {target}");
                    }
                    return "Applying patch".to_string();
                }
                if let Some(summary) = mcp_tool_summary(&tool_name, tool_input) {
                    return summary;
                }
            }
            format!("Using {tool_name}")
        }
        "PostToolUse" => {
            if let Some(content) = event
                .payload
                .get("content")
                .and_then(serde_json::Value::as_str)
                .and_then(content_preview)
            {
                if tool_name == "exec_command" {
                    return format!("Command completed: {}", truncate(&content, 72));
                }
                return format!("{tool_name} completed: {}", truncate(&content, 72));
            }
            format!("{tool_name} completed")
        }
        "PostToolUseFailure" => {
            if let Some(content) = event
                .payload
                .get("content")
                .and_then(serde_json::Value::as_str)
                .and_then(content_preview)
            {
                return format!("{tool_name} failed: {}", truncate(&content, 72));
            }
            format!("{tool_name} failed")
        }
        "SessionStart" => payload_string(&event.payload, "title")
            .map(|title| format!("Watching {}", truncate(&title, 80)))
            .unwrap_or_else(|| "Session observed".to_string()),
        "SessionEnd" => "Session ended".to_string(),
        "SubagentStart" => {
            let label = payload_string(&event.payload, "display_name")
                .or_else(|| payload_string(&event.payload, "agent_name"))
                .or_else(|| payload_string(&event.payload, "agent_type"))
                .or_else(|| event.agent_name.clone())
                .unwrap_or_else(|| "Agent".to_string());
            if let Some(description) = payload_string(&event.payload, "description") {
                return format!("Spawned {} to {}", label, truncate(&description, 72));
            }
            format!("Spawned {label}")
        }
        "SubagentStop" => "Subagent finished".to_string(),
        "UserPromptSubmit" => payload_string(&event.payload, "prompt")
            .or_else(|| payload_string(&event.payload, "message"))
            .and_then(|prompt| clean_summary_text(&prompt))
            .map(|prompt| format!("Prompted: {}", truncate(&prompt, 72)))
            .unwrap_or_else(|| "User prompt".to_string()),
        "AssistantResponse" => payload_string(&event.payload, "text")
            .and_then(|text| clean_summary_text(&text))
            .map(|text| format!("Responded: {}", truncate(&text, 72)))
            .unwrap_or_else(|| "Response".to_string()),
        "SessionTitleChanged" => {
            payload_string(&event.payload, "title").unwrap_or_else(|| "Title changed".to_string())
        }
        _ => event.hook_event_type.clone(),
    }
}

pub(crate) fn extract_command(tool_input: &serde_json::Value) -> Option<String> {
    if let Some(command) = tool_input.get("cmd").and_then(serde_json::Value::as_str) {
        let trimmed = command.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    if let Some(command) = tool_input
        .get("command")
        .and_then(serde_json::Value::as_str)
    {
        let trimmed = command.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    if let Some(parts) = tool_input
        .get("command")
        .and_then(serde_json::Value::as_array)
    {
        let joined = parts
            .iter()
            .filter_map(serde_json::Value::as_str)
            .map(str::trim)
            .filter(|part| !part.is_empty())
            .collect::<Vec<_>>()
            .join(" ");
        if !joined.is_empty() {
            return Some(joined);
        }
    }
    None
}

pub(crate) fn extract_file_target(tool_input: &serde_json::Value) -> Option<String> {
    ["file_path", "path", "file"]
        .iter()
        .find_map(|key| tool_input.get(*key).and_then(serde_json::Value::as_str))
        .and_then(short_path)
}

pub(crate) fn extract_patched_file(patch: &str) -> Option<String> {
    let marker = ["*** Add File: ", "*** Update File: ", "*** Delete File: "];
    for prefix in marker {
        if let Some(rest) = patch.lines().find_map(|line| line.strip_prefix(prefix)) {
            return short_path(rest);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use crate::model::LegacyHookEvent;

    use super::describe_legacy_event;

    fn build_event(payload: serde_json::Value) -> LegacyHookEvent {
        LegacyHookEvent {
            source_app: "pharos".to_string(),
            session_id: "session-1".to_string(),
            hook_event_type: "PreToolUse".to_string(),
            payload,
            timestamp: 1,
            agent_id: None,
            agent_type: None,
            model_name: None,
            display_name: None,
            agent_name: None,
        }
    }

    #[test]
    fn describes_memory_brain_mcp_calls() {
        let event = build_event(json!({
            "tool_name": "CallMcpTool",
            "tool_input": {
                "server": "ai-memory-brain",
                "toolName": "memory_store_summary"
            }
        }));
        assert_eq!(describe_legacy_event(&event), "Memory brain: store summary");
    }

    #[test]
    fn describes_fetch_mcp_resource_calls() {
        let event = build_event(json!({
            "tool_name": "FetchMcpResource",
            "tool_input": {
                "server": "ai-memory-brain",
                "uri": "memory://brain/health"
            }
        }));
        assert_eq!(
            describe_legacy_event(&event),
            "Fetching MCP resource ai-memory-brain: memory://brain/health"
        );
    }
}
